0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+  * The most common base case is an expression of the form `x in C`, where `C` is an expression with a sequence, multiset, or (i)set type. These expressions produce enumerations that bind only `x` to each value in the collection.
+    * For sequences, the ordering is the obvious ordering of values in the sequence.
+    * For (i)sets or multisets, the exact ordering is deterministic but unspecified.
+    * Potentially-infinite sets are only allowed if they are enumerable.
+    * If `x` is not a variable, then the enumeration produces an empty set of bindings once for each time the LHS appears in the RHS collection.
+  * For an expression of the form `A && B`, the enumeration includes all bindings that satisfy both `A` and `B`. The enumeration ordering is defined by ordering first by `A`'s ordering and then by `B`'s, and the multiplicities of bindings is multiplied. See below for the pseudocode that calculates this enumeration.
+    * This means that the `&&` operation is not fully symmetrical, but this is already true as verification considers it to be "short-circuiting": 
+      `b != 0 && a/b > 1` is valid but `a/b > 1 && b != 0` is not.
+    * This definition is chosen such that the multiplicities of results from `A && B` is the same as `B && A`, even if the ordering is not the same.
+  * For an expression of the form `A || B`, the enumeration will simply produce the values enumerated by `A` followed by those enumerated by `B`.
+  * If the context of an expression binds a variable that is not used in the expression, such as `seq x: bool | true`, 
+    then the default ordering for the type of the variable is used (i.e. `[false, true]` in this case). 
+    If this type is not enumerable, then the expression is not allowed.
+  * For an expression of the form `!A`, the enumeration will produce all values of the types of the bound variables that do not satisfy `A`. 
+    As above, if any of these types are not enumerable the expression is not allowed.
+
+Here is the pseudocode for the semantics of the ordered enumeration of bindings for `A && B`:
+
+```
+IEnumerable<ActualBindings> EnumerateConjunction(Expr A, Expr B) {
+  foreach (ActualBindings a in A) {
+    Expr B' := B[a];  // Substitute the given bindings into the expression B
+    foreach (ActualBindings b' in B') {
+      yield return b';
+    }
+  }
+}
+```
+
+And some illustrative examples using sequence comprehensions:
+
+```dafny
+// Special cases for one variable:
+seq x: T | A && B == seq x | A * seq x | B  // Not an actual built-in operation, but a generalization of set intersection.
+seq x: T | A || B == seq x | A + seq x | B
+seq x: T | !A == seq x: T - seq x | A       // Not an actual built-in operation, but a generalization of set subtraction.
+                                            // Only well-formed if T itself is enumerable.
+
+seq x | x in [1, 2, 2] && x in [2, 2, 1] == [1, 2, 2, 2, 2]
```

Comment:
Why not 
```suggestion
seq x | x in [1, 2, 2] && x in [2, 2, 1] == [1, 2, 2, 2, 2, 1]
```
?
Also:
```suggestion
(seq x | x in [1, 2, 2] && x in [2, 2, 1]) == [1, 2, 2, 2, 2]
```
If you could fix parentheses below that'd be great. Parsing will always consider the seq expression to continue.

Reply:
Oh interesting. It's not obvious at all but it makes sense now. That would work the same way for a multiset, right?